### PR TITLE
Enable support for exporting 1D and 2D dataset slices.

### LINKED ISSUE #20: Exporting slices
Hi all,

I am using the H5Web extension in VSCode and trying to use the export functionality; I want to save certain slices to another files. However the export button/functionality is always disabled. Is there something I need to do to enable it?

Thank you!

### Is your feature request related to a problem?
I cannot export slices to CSV. The "Export" bottom is disabled.

### Alternatives you've considered

Copy/paste

### Additional context
Screen Shot 2022-12-23 at 11 39 15 AM

### PR DESCRIPTION
**Pull request by mandrew9**
Fix #20 

I'm admittedly not an expert with TypeScript or VS Code extensions, but I think this implementation for exporting data slices within the VS Code extension follows your guidelines provided in the associated ticket. Please let me know if there's anything I can do to improve the implementation.

### CODE REVIEW

File: app/Viewer.tsx [hunk 1]
```diff
@@ -23,7 +24,65 @@ function Viewer(props: Props) {
...
+                  csv += `${value[k].toString()}\n`
+                  k++
+                }
+              } else {
+                value.forEach((val) => {
+                  csv += `${val.toString()}\n`
+                })
+              }
+
+              const finalCsv = csv.slice(0, -1);
```
Comment (axelboc) [inline app/Viewer.tsx:71]: With my suggestion above, I believe `-1` would have to be replaced with `-2` to remove both the last line break and last coma.

**Review by axelboc — CHANGES_REQUESTED** [review]
Thanks for having a crack at this @mandrew9!

File: app/Viewer.tsx [hunk 2]
```diff
@@ -23,7 +24,65 @@ function Viewer(props: Props) {
...
+              let dims = [];
+
+              // Record all dataset dimensions with cardinality greater than one,
+              // accounting for subset selection if applicable
+              if (selection) {
+                let subsets = selection.split(',')
+                for (let i = 0; i < subsets.length; i++) {
+                  // Because of the user interface, all slices will either be
+                  // the full array length or a single value
+                  if (subsets[i] === ':' && dataset.shape[i] > 1) {
```
[the remark this answers is not in the document]
Comment (mandrew9) [reply app/Viewer.tsx:43]: Sorry it took me much longer to get back to this than I expected. Regarding this issue with transposition of the data, I think the problem is more complex. Based on the testing that I did, the user may transpose the data in the UI, but those changes are not reflected in the data about dimensionality sent to this function. Do you know if there is a means available to recover the dimensionality of the data as currently being viewed within the UI?

[the remark this answers is not in the document]
Comment (axelboc) [reply app/Viewer.tsx:43]: Hmm, indeed you're right! I'll get back to you on this.

[the remark this answers is not in the document]
Comment (axelboc) [reply app/Viewer.tsx:43]: Alright, I've opened an issue on the main H5Web repo: https://github.com/silx-kit/h5web/issues/1454 -- but it's not a blocking issue. Let's accept potential transpositions in the export for now.

File: app/App.tsx [hunk 3]
```diff
@@ -7,6 +7,7 @@ import { ErrorBoundary } from 'react-error-boundary';
 
 function App() {
   const [fileInfo, setFileInfo] = useState<FileInfo>();
+  const [vscode, setVsCodeApi] = useState<VsCodeApi>();
```
[the remark this answers is not in the document]
Comment (axelboc) [reply app/App.tsx]: In #29, I moved the `acquireVsCodeApi` call to a dedicated file – no need for a `useEffect` at all in the end.

File: app/Viewer.tsx [hunk 4]
```diff
@@ -23,7 +24,65 @@ function Viewer(props: Props) {
   }, [fileInfo]);
 
   return (
-    <H5WasmProvider filename={fileInfo.name} buffer={buffer}>
+    <H5WasmProvider filename={fileInfo.name} buffer={buffer}
+        getExportURL={(format, dataset, selection, value) => {
+          if (format === 'csv') {
```
[the remark this answers is not in the document]
Comment (axelboc) [reply app/Viewer.tsx]: No longer relevant now that we also have the JSON export.

**Comment by axelboc** [comment]
I've refactored the CSV export logic a bit to avoid filtering out dimensions with cardinality of 1 and to avoid having a specific `for` loop for 1D datasets/slices. I've also removed the extra space that was added after every comma.

Thanks again @mandrew9!

**Review by axelboc — APPROVED** [review]

**Pull request merged**